relays/ethereum/src/substrate_client.rs

```diff
@@ -56,23 +102,42 @@ impl MaybeConnectionError for Error {
 }
 
 /// Returns client that is able to call RPCs on Substrate node.
-pub fn client(uri: &str) -> Client {
-	let transport = HttpTransportClient::new(uri);
+pub fn client(params: SubstrateConnectionParams) -> Client {
+	let uri = format!("http://{}:{}", params.host, params.port);
+	let transport = HttpTransportClient::new(&uri);
 	Client {
 		rpc_client: RawClient::new(transport),
 		genesis_hash: None,
 	}
 }
 
+/// Returns best Substrate header.
+pub async fn best_header(client: Client) -> (Client, Result<SubstrateHeader, Error>) {
+	call_rpc(client, "chain_getHeader", Params::None, rpc_returns_value).await
+}
+
+/// Returns Substrate header by hash.
+pub async fn header_by_hash(client: Client, hash: Hash) -> (Client, Result<SubstrateHeader, Error>) {
+	let hash = bail_on_arg_error!(to_value(hash).map_err(|e| Error::RequestSerialization(e)), client);
+	call_rpc(client, "chain_getHeader", Params::Array(vec![hash]), rpc_returns_value).await
+}
+
+/// Returns Substrate header by number.
+pub async fn header_by_number(client: Client, number: Number) -> (Client, Result<SubstrateHeader, Error>) {
+	let (client, hash) = bail_on_error!(block_hash_by_number(client, number).await);
+	header_by_hash(client, hash).await
+}
+
 /// Returns best Ethereum block that Substrate runtime knows of.
 pub async fn best_ethereum_block(client: Client) -> (Client, Result<EthereumHeaderId, Error>) {
-	let (client, result) = call_rpc::<(u64, H256)>(
+	let (client, result) = call_rpc(
```

Comment:
Same as wtih ethereum_client, can we extract this low-level calls (including serialization to some encapsulated module)? IMHO it would be way more readable if we avoided mixing bridge logic with rpc payload boilerplate.

Reply:
Logger here: https://github.com/paritytech/parity-bridges-common/issues/74